### PR TITLE
Add Solar Energy models

### PR DESCRIPTION
**Pull request by stevenkj1**
Add Solar Models to publicly available examples.

### CODE REVIEW

File: SolarEnergy/Aero/aero.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                        "ft22m2": 0.092903, "m22ft2": 10.7639,
+                        "kgpm32slugspft3": 0.00194032, "slugspft32kgpm3": 515.37891,
+                        "kgpm32lbpft3": 0.062428, "lbpft32kgpm3": 16.01}
+        self.startdatetime = None
+
+    def Move(self, ctarget):
+        # Note fix the speed to get this working in DAFNI, easy to change later.
+        self.CheckSpeed(h=self.position.alt)
+        # Get the power (irad) for this position
+        #print("Currently at : ", self.position.lon, self.position.lat, "(travelling to ", ctarget.lon, ctarget.lat, ")")
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Commented out print statement. Can remove if we aren't using this for logging.

File: SolarEnergy/Aero/aero.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            sthtl.append(self.AirDensity(height)*self.convert.get('slugspft32kgpm3'))   # Convert dens. to SI for output.
+            sthtl.append(self.T(h=height)-273.15)     # 273.15 converts to Centigrade
+            sthtl.append(self.GetStall(height))
+            sthtl.append(self.GetMaxVelocityBounds(scl=self.maxpower, h=height))
+            tspdl = []
+            for spd in speedl:
+                tmpl = []
+                tmpl.append(spd)
+                tmpl.append(self.maxpower)
+                tmpl.append(self.PowerReqHP(Vf=spd*self.convert.get('mph2fps'), h=height)) # KJS wip
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Work in progress??

File: SolarEnergy/Aero/aero.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        '''Returns the aspect ratio = Wingspan**2/Wing area (both are in ft)'''
+        self.AR = self.b**2/self.S
+        print("Aspect Ratio is ", self.AR)
+
+    def _GVB_ftn(self, Vf, scl, h=0.0):
+        '''Internal function used to evaluate roots of the power curve (Newton-Raphson method)'''
+        '''Units (mix of Imp. & SI): velocity in fps, scl (power) in hp, h(eight) in m'''
+        return self.PowerReqHP(Vf, h)-scl
+
+    def GetMaxVelocityBounds(self, scl=70.0, h=0.0): # Fix height dependence donem scl is max power ... neaten this
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Typo in comment

File: SolarEnergy/Aero/aero.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        '''Returns the maximum speed in mph (from the power-curve). scl is the max engine power (hp), h in meters'''
+        # P(vf) - scl = PowerReqHP(self, Vf) -scl  # hcl & h are fixed params in the root finding
+        v2 = optimize.newton(self._GVB_ftn, 100.0, args=(scl, h,))  # 100.0 is a start estimate for the root.
+        print("v2 = ", v2*self.convert.get('fps2mph')) # 0.681 is fts-1 -> mph
+        return v2*self.convert.get('fps2mph')
+
+    def GetStall(self, h=0.0):
+        '''Returns the stall speed in mph. height in metres'''
+        vs = math.sqrt(2.0*self.weight/(self.AirDensity(h)*self.S*1.5)) # CL_max to 1.5 ? (1-2,  with flaps - no idea ?)
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: No idea....??

File: SolarEnergy/Aero/aero.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        return self.PowerReq(Vf, h)*1.36 # Feet.lb/s -> Watts
+
+    def PowerReqHP(self, Vf, h=0.0):
+        '''Evaluates the power(hp) needed for level flight, given velocity-TAS(fps) and height(m).'''
+        return self.PowerReq(Vf, h)/550.0 # Feet.lb/s -> hp
+
+    def CL(self, Vf, h=0.0):
+        '''Lift coefficient given vf(fps) and h(m)'''
+        CL = self.weight / (0.5 * self.AirDensity(h) * self.S * Vf**2)
+        #print("CL = ", CL)
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Again, commented out print statement,

File: SolarEnergy/Aero/main_avi.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    else:
+        parser = argparse.ArgumentParser()
+        parser.add_argument('--startdaytime', type=str)
+        parser.add_argument('--speed', type=float)
+        args = parser.parse_args()
+        inp = Inputs(args)
+
+    xfp, irTest = loaddata()
+    # We want to use the earlist time from the index of the irradiance data, rather than setting it twice.
+    # print("Get any index : ", irTest.dflist[0].index[0], type(irTest.dflist[0].index[0]))
```
Comment (estherturner) [inline SolarEnergy/Aero/main_avi.py]: And again.

File: SolarEnergy/Aero/waypoints.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    def __init__(self, iniday=None, initime=None, inipos=None):
+        self.iniday = iniday
+        self.initime = initime
+        self.inipos = inipos
+        self.waypoints = [inipos]
+        self.iwaypoints = None
+
+    def GenIData(self):
+        # Mark original waypoints as 1 (at some point I think it'll be best to redo this a bit so it's less clunky)
```
Comment (estherturner) [inline SolarEnergy/Aero/waypoints.py]: I appreciate you probably still want to redo this, but maybe it shouldn't be noted in a code comment if we're making this public. Or at least, word it a bit differently.

File: SolarEnergy/Aero/waypoints.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            print(wapn[0])
+            self.iwaypoints.append(wapn[0])
+            self.__GenSubpoints(wp1=wapn[0], wp2=wapn[1])
+        self.iwaypoints.append(self.waypoints[-1])
+
+    def __GenSubpoints(self, wp1=None, wp2=None):
+        # Generate wp1 Ir.data
+        # Generate set of points between waypoints
+        # i) Gen. geodisic (shortest path on ellipsoid) connecting wp1 & wp2
+        gds = Geodesic.WGS84.Inverse(wp1.lat, wp1.lon, wp2.lat, wp2.lon ) # Arg.. bloody typo.. wp1 was usedx3 !!
```
Comment (estherturner) [inline SolarEnergy/Aero/waypoints.py]: Probably should remove this comment!

File: SolarEnergy/Aero/waypoints.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        for i in range(splt):
+            # Project 20km towards wp2 & create an intermediate wp. (rem ip is a geopy POINT, not my Waypoint class)
+            ip = mdist.destination((iwp.lat, iwp.lon), bearing=gds['azi1'])
+            iwp = Waypoint(lat=ip.latitude, lon=ip.longitude, alt=wp1.alt, speed=wp1.speed)
+            self.iwaypoints.append(iwp)
+
+    def PrintPoints(self):
+        #print("Printing Main Waypoints")
+        #for i, waypoint in enumerate(self.waypoints):
+        #    print("point", i, waypoint.lon, waypoint.lat, waypoint.alt)
```
Comment (estherturner) [inline SolarEnergy/Aero/waypoints.py]: Again, can remove these print statements,

File: SolarEnergy/Aero/waypoints.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        print("Printing Full Set of Waypoints (inc. intermediates)")
+        for i, iwaypoint in enumerate(self.iwaypoints):
+            print("point", i, iwaypoint.lon, iwaypoint.lat, iwaypoint.alt, iwaypoint.speed, iwaypoint.id)
+
+    def plotcourse(self):
+        #import plotly.express as plx  # Need to pip this & create a flight-path dataframe for plotting.
+        #fig = plx.line_mapbox(self.fpdf, lat="lat", lon="lon", color="State", zoom=3, height=300)
+        #fig.update_layout(mapbox_style="stamen-terrain", mapbox_zoom=2, mapbox_center_lat = 25.0, 
+        #                  margin={"r":0,"t":0,"l":0,"b":0})
+        #fig.show()
```
Comment (estherturner) [inline SolarEnergy/Aero/waypoints.py]: Is there a reason this plotting code is still here but not used and that the function just passes?

File: SolarEnergy/Aero/waypoints.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+            tmpwp = Waypoint(lon=float(ldata[0]), lat=float(ldata[1]), alt=float(ldata[2]), 
+                             speed=float(ldata[3]), id=int(ldata[4]))
+            self.iwaypoints.append(tmpwp)
+            if tmpwp.id == 1:
+                self.waypoints.append(tmpwp)
+        # Note GenIData will create track of intermediate waypoints. # Naahh better to do this manually !
+        #if len(self.waypoints) == len(self.iwaypoints):
+        #    self.GenIData()
+
+    #def LoadIrData
```
Comment (estherturner) [inline SolarEnergy/Aero/waypoints.py]: This looks like working, get rid of it?

File: SolarEnergy/Irradiance/irradiance.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        if not HASMPL:
+            print("MatPlotlib not installed : skipping graphs")
+            return
+        self.poa_df_dy = self.poa_df.copy()
+        self.poa_df_dy2 = self.poa_df.copy()
+        # Winter
+        self.poa_df_dy['Date'] = self.poa_df_dy.index
+        self.poa_df_dy = self.poa_df_dy[(self.poa_df_dy['Date'] > '2023-01-01 00:00') & (self.poa_df_dy['Date'] < '2023-01-01 23:59')]
+        self.poa_df_dy.index = self.poa_df_dy.index.strftime("%H:%M")
+        # Summer (surely it's possible to do this in pandas without all the copy stuff !!)
```
Comment (estherturner) [inline SolarEnergy/Irradiance/irradiance.py]: Neaten up this comment.

File: SolarEnergy/Irradiance/irradiance.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            file = self.poa_df.to_csv(filep)
+        elif ext == ".pqt" or ext == ".parquet":
+            file = self.poa_df.to_parquet(filep)
+
+
+class IrData():
+
+    def __init__(self):
+        self.waypoint = None
+        self.Idataframe = None  # Dont use this do I ? Yeah just use a generic zip below
```
Comment (estherturner) [inline SolarEnergy/Irradiance/irradiance.py]: Again, neaten this up.

File: SolarEnergy/Irradiance/main_solar.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            iddfp = glob.glob(spath)
+            if len(iddfp) == 1:
+                datafp = os.path.join(iddfp[0])
+            else:
+                datafp = os.path.join(os.path.join(gPATHI, "waypoints.csv"))
+        else:
+            datafp = os.path.join(os.path.join(gPATHI, inp.inFile))
+        print("Loading journey from Waypoint file, with a start time of", inp.StartDayTime)
+        xfp = FlightPath(iniday=inp.StartDayTime)
+        xfp.LoadIData(datafp) # Change this to be just the main points. Then do whatever path.py does !
```
Comment (estherturner) [inline SolarEnergy/Irradiance/main_solar.py]: Is this a note for yourself?

**Review by estherturner — COMMENTED** [review]
I've skim read this very very quickly. Which there are definitely changes that would be good to make in terms of coding style/standards, I've ignored this. I also haven't checked the functionality at all. I've mainly gone through checking the comments (because if I look at a piece of code for the first time, that's what I'll read). There are a few bits of commented out code that I'm wondering if we can remove to neaten things up, and then some explanatory comments that I think we should do some tidying up of.

File: SolarEnergy/Aero/aero.py (hunk 2 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Old comment, removed.

File: SolarEnergy/Aero/aero.py (hunk 3 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Removed this one as well

File: SolarEnergy/Aero/aero.py (hunk 1 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Put in an : 
if gDEBUG: 
     print .....
here

File: SolarEnergy/Aero/aero.py (hunk 4 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Long version :  CL_max is the maximum lift coefficient - which defines when a wing stalls. Short of a wind-tunnel or running fluid-dynamics sims I've not idea what exactly this is.
Short version : Rough guess probably isn't too far off the mark for this plane & comment edited :-)  !

File: SolarEnergy/Aero/aero.py (hunk 5 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: removed print

File: SolarEnergy/Aero/main_avi.py (hunk 6 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/main_avi.py]: removed print

File: SolarEnergy/Irradiance/main_solar.py (hunk 14 above)
Comment (stevenkj1) [reply SolarEnergy/Irradiance/main_solar.py]: comment removed

File: SolarEnergy/Irradiance/irradiance.py (hunk 13 above)
Comment (stevenkj1) [reply SolarEnergy/Irradiance/irradiance.py]: Removed comment

File: SolarEnergy/Irradiance/irradiance.py (hunk 12 above)
Comment (stevenkj1) [reply SolarEnergy/Irradiance/irradiance.py]: removed the comment

File: SolarEnergy/Aero/CbData.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    def GetNDF(self, position): # Given a time & a place give the POA (plane of array irradiance)
+        # Loop to get closest position (again this can be shortend to speed things up if needed)
+        from geographiclib.geodesic import Geodesic # move up to top later
+        closep = 26000.0 # km
+        closed = None
+        for data in self.clist:
+            wpdat = data[0]
+            gds = Geodesic.WGS84.Inverse(position.lat, position.lon, wpdat.lat, wpdat.lon )
+            dpts = gds['s12']/1000.0  # distance between points. Work in km.
+            #print("distance", dpts)
```
Comment (estherturner) [inline SolarEnergy/Aero/CbData.py]: Sorry, missed this before!

File: SolarEnergy/Aero/CbData.py [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+            if dpts < closep:
+                closep = dpts
+                closed = data
+        # Really need to check this is working - is it ok !?
```
Comment (estherturner) [inline SolarEnergy/Aero/CbData.py]: Again, sorry, missed this before!

File: SolarEnergy/Aero/CbData.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+    def GetIrradiance(self, position=None, datetime=None):
+        '''Takes a waypoint object & a datetime string'''
+        ndf = self.GetNDF(position)
+        # Get the right data using the datetime obj, ie. nearest index to datetime
+        dt = pd.to_datetime(datetime)
+        irow = ndf[1].index.get_indexer([dt], method='nearest')[0]
+        irrval = ndf[1].iat[irow, 0]
+        print("Getting Irradiance :", irrval, "at", datetime)
+        return irrval # was 800 # The time bit is working ok I reckon !!
```
Comment (estherturner) [inline SolarEnergy/Aero/CbData.py]: And this... What was I doing the last time I read this?

File: SolarEnergy/Aero/aero.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+        # Engine related
+        self.maxpower = 70.0
+        self.maxbattpack = 633.0*260.0 * 3600.0   # BattWeight(kg)*Density, in Watt-hrs, conv. to Joules with 3600 (60x60)
+        self.currentbatt = self.maxbattpack # Start at full battery power
+        self.egain = 0.0 # Energy loss & gain in last time increment.
+        self.eloss = 0.0
+        self.irrad = 0.0 # Irradiance for iteration
+        self.record = []
+        self.recordPD = []
+        # Constants (propeff will depend on height - ie. airdensity, to a 1/3 power I think ?)
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Can we get rid of the 'I think?'

File: SolarEnergy/Aero/aero.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+        R = 287.0 # J/kgK
+        alpha = -1.0*(g0/(a*R) + 1.0)
+        T = self.T(T0, a, h)
+        rho = rho0*(T/T0)**alpha
+        if gDEBUG:
+            print("Density = ", rho, "@", h, "m", " T=", T)
+        density = rho*self.convert.get("kgpm32slugspft3")
+        return density
+    
+    def T(self, T0=288.16, a=-0.0065, h=0): # fix this properly @ not 8am. KJS
```
Comment (estherturner) [inline SolarEnergy/Aero/aero.py]: Deal with this comment.

**Review by estherturner — COMMENTED** [review]
Sorry, found a few more comments I think we should probably deal with! But thanks very much for the previous iteration!

File: SolarEnergy/Aero/CbData.py (hunk 15 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/CbData.py]: Fixed.

File: SolarEnergy/Aero/CbData.py (hunk 17 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/CbData.py]: Removed

File: SolarEnergy/Aero/aero.py (hunk 19 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Removed this one.

File: SolarEnergy/Aero/aero.py (hunk 18 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/aero.py]: Removed this comment, it depends on the propeller. a fixed 85% is going to be close enough for these purposes.

File: SolarEnergy/Aero/CbData.py (hunk 16 above)
Comment (stevenkj1) [reply SolarEnergy/Aero/CbData.py]: Removed

**Pull request merged**